v-next/hardhat-viem/src/chains.ts

```diff
@@ -0,0 +1,139 @@
+import type { TestClientMode } from "./types.js";
+import type { EthereumProvider } from "@ignored/hardhat-vnext/types/providers";
+import type * as viemT from "viem";
+
+import { HardhatError } from "@ignored/hardhat-vnext-errors";
+import { extractChain } from "viem";
+import * as chainsModule from "viem/chains";
+import { hardhat, anvil } from "viem/chains";
+
+/* eslint-disable-next-line @typescript-eslint/consistent-type-assertions
+-- TODO: this assertion should not be necessary */
+const chains = Object.values(chainsModule) as viemT.Chain[];
+
+export async function getChain(
+  provider: EthereumProvider,
+): Promise<viemT.Chain> {
+  const chainId = await getChainId(provider);
+
+  const chain = extractChain({
+    chains,
+    id: chainId,
+  });
+
+  if (isDevelopmentNetwork(chainId) || chain === undefined) {
+    if (await isHardhatNetwork(provider)) {
+      return {
+        ...hardhat,
+        id: chainId,
+      };
+    }
+
+    if (await isAnvilNetwork(provider)) {
+      return {
+        ...anvil,
+        id: chainId,
+      };
+    }
+
+    // If the chain couldn't be found and we can't detect the development
+    // network we throw an error
+    if (chain === undefined) {
+      throw new HardhatError(HardhatError.ERRORS.VIEM.NETWORK_NOT_FOUND, {
+        chainId,
+      });
+    }
+
+    // If the chain is a development network but we couldn't detect which one
+    // we throw an error
+    throw new HardhatError(
+      HardhatError.ERRORS.VIEM.UNSUPPORTED_DEVELOPMENT_NETWORK,
+    );
+  }
+
+  return chain;
+}
+
+const chainIdCache = new Map<EthereumProvider, number>();
+
+export async function getChainId(provider: EthereumProvider): Promise<number> {
+  const cachedChainId = chainIdCache.get(provider);
+  if (cachedChainId !== undefined) {
+    return cachedChainId;
+  }
+
+  const chainId = Number(await provider.request({ method: "eth_chainId" }));
+  chainIdCache.set(provider, chainId);
+
+  return chainId;
+}
+
+export function isDevelopmentNetwork(chainId: number): boolean {
+  return chainId === 31337;
+}
+
+enum NetworkMethod {
+  HARDHAT_METADATA = "hardhat_metadata",
+  ANVIL_NODE_INFO = "anvil_nodeInfo",
+}
+
+const isHardhatNetworkCache = new Map<EthereumProvider, boolean>();
+
+export async function isHardhatNetwork(
+  provider: EthereumProvider,
+): Promise<boolean> {
+  const cachedIsHardhat = isHardhatNetworkCache.get(provider);
+  if (cachedIsHardhat !== undefined) {
+    return cachedIsHardhat;
+  }
+
+  const isHardhat = await isMethodSupported(
+    provider,
+    NetworkMethod.HARDHAT_METADATA,
+  );
+  isHardhatNetworkCache.set(provider, isHardhat);
+
+  return isHardhat;
+}
+
+const isAnvilNetworkCache = new Map<EthereumProvider, boolean>();
+
+export async function isAnvilNetwork(
```

Comment:
This was previously called `isFoundryNetwork`, but since the `foundry` chain has been marked as deprecated in the `viem/chains` module (and `anvil` is now recommended), I renamed it accordingly.